File: Itido-BMS/dart/bms_dart/lib/src/models/request.dart
import 'package:bms_dart/src/models/approval_state.dart';

import 'date_time_converter.dart';
import 'user.dart';

class Request {
  ApprovalState approvalState;
  bool canRespondToApprovalState;
  DateTime respondDateTime;
  User responder;

  Request({
    this.approvalState,
    this.canRespondToApprovalState,
    this.respondDateTime,
    this.responder,
  });

  factory Request.fromJson(json, {rootJson, path}) {
    if (json == null) return null;
    return Request._fromJson(json, rootJson ?? {}, path ?? '');
  }

  Request._fromJson(json, rootJson, path)
      : this.approvalState = ApprovalState.values[json['approvalState']],
        this.canRespondToApprovalState = json['canRespondToApprovalState'],
        this.respondDateTime = toDateTime(json['respondDateTime']),
        this.responder = User.fromJson(
          json['responder'] ?? rootJson['${path}_responder'],
        );
}

File: Itido-BMS/RentAPI/Rent/Models/TimePlanning/Request.cs
using System;
using System.Linq.Expressions;

namespace Rent.Models.TimePlanning
{
    public class Request
    {
        public int ID { get; set; }
        public ApprovalState ApprovalState { get; set; }
        public DateTime? RespondDateTime { get; set; }

        public int? CreatorID { get; set; }
        public virtual User Creator { get; set; }

        public int? SubjectID { get; set; }
        public virtual User Subject { get; set; }

        public int? ResponderID { get; set; }
        public virtual User Responder { get; set; }

        public static Expression<Func<Request, User, dynamic>> BasicDTO(int requester, string requesterRole)
        {
            var isRequesterAdminOrManager = requesterRole.Equals("Admin") || requesterRole.Equals("Manager");

            return (request, creator) => request != null ?
            new
            {
                request.ID,
                request.ApprovalState,
                //CanRespondToApprovalState = false,
                CanRespondToApprovalState = request.ApprovalState == ApprovalState.Pending &&
                        (((creator.UserRole.Equals("Admin") || creator.UserRole.Equals("Manager")) && (request.SubjectID == requester) ||
                        (isRequesterAdminOrManager && request.SubjectID == request.CreatorID))),
            } : null;
        }
    }
}
